ext/standard/var.c

```diff
@@ -868,6 +868,11 @@ static int php_var_serialize_get_sleep_props(
 		zend_string *name, *tmp_name, *priv_name, *prot_name;
 
 		ZVAL_DEREF(name_val);
+
+		if (Z_TYPE_P(name_val) == IS_INDIRECT) {
+			ZVAL_DEINDIRECT(name_val);
+		}
```

Comment:
This should be before the DEREF, as it could be both indirect and ref. Though really, just use ZEND_HASH_FOREACH_VAL_IND.

The question I'm really asking myself though is why we're accepting an object as the return value from `__sleep` in the first place. `__sleep` is supposed to return an array of names, and object doesn't really make sense in that context. I suspect this is just overeager use of HASH_OF. Note also how we reject non-array types on `__sleep`: https://3v4l.org/Hq5ei

I think it would be better to drop the object support on master than do this change.

Reply:
Ok, new approach, how wrong is this one ? :)